cirq/experiments/xeb_fitting.py

```diff
@@ -289,10 +306,209 @@ def _mean_infidelity(angles):
+        exponential_decay, cycle_depths, fidelities, p0=(A_0, layer_fid_0), bounds=((0, 0), (1, 1))
+    )
+    return A, layer_fid
+
+
+def _one_unique(df, name, default):
+    """Helper function to assert that there's one unique value in a column and return it."""
+    if name not in df.columns:
+        return default
+    vals = df[name].unique()
+    assert len(vals) == 1, name
+    return vals[0]
+
+
+def fit_exponential_decays(fidelities_df: pd.DataFrame) -> pd.DataFrame:
+    """Fit exponential decay curves to a fidelities DataFrame.
+
+    Args:
+         fidelities_df: A DataFrame that is the result of `benchmark_2q_xeb_fidelities`. It
+            may contain results for multiple pairs of qubits identified by the "pair" column.
+            Each pair will be fit separately. At minimum, this dataframe must contain
+            "cycle_depth", "fidelity", and "pair" columns.
+
+    Returns:
+        A new, aggregated dataframe with index given by (pair, layer_i, pair_i); columns
+        for the fit parameters "A" and "layer_fid"; and nested "cycles_depths" and "fidelities"
+        lists (now grouped by pair).
+    """
+    records = []
+    for pair in fidelities_df['pair'].unique():
+        f1 = fidelities_df[fidelities_df['pair'] == pair]
+        A, layer_fid = _fit_exponential_decay(f1['cycle_depth'], f1['fidelity'])
+        record = {
+            'pair': pair,
+            'A': A,
+            'layer_fid': layer_fid,
+            'cycle_depths': f1['cycle_depth'].values,
+            'fidelities': f1['fidelity'].values,
+            'layer_i': _one_unique(f1, 'layer_i', default=0),
+            'pair_i': _one_unique(f1, 'pair_i', default=0),
+        }
+        records.append(record)
+    return pd.DataFrame(records).set_index(['pair', 'layer_i', 'pair_i'])
+
+
+def before_and_after_characterization(
+    fidelities_df_0: pd.DataFrame, characterization_result: XEBCharacterizationResult
+) -> pd.DataFrame:
+    """A convenience function for horizontally stacking results pre- and post- characterization
+    optimization.
+
+    Args:
+        fidelities_df_0: A dataframe (before fitting), likely resulting from
+            `benchmark_2q_xeb_fidelities`.
+        characterization_result: The result of running a characterization. This contains the
+            second fidelities dataframe as well as the new parameters.
+
+    Returns:
+          A joined dataframe with original column names suffixed by "_0" and characterized
+          column names suffixed by "_c".
+    """
+    fit_decay_df_0 = fit_exponential_decays(fidelities_df_0)
+    fit_decay_df_c = fit_exponential_decays(characterization_result.fidelities_df)
+
+    joined_df = fit_decay_df_0.join(fit_decay_df_c, how='outer', lsuffix='_0', rsuffix='_c')
+    joined_df['characterized_angles'] = [
+        characterization_result.final_params[pair] for pair, _, _ in joined_df.index
+    ]
+    angle_names = list(list(characterization_result.final_params.values())[0].keys())
```

Comment:
Why double list()?

Reply:
I suppose the outer list isn't necessary. The inner one is because I want the first value of `characterization_result.final_params` and dictionaries return iterators which don't allow indexing.

I've re-worked this using fancy unpacking to avoid *any* `list` calls. How about that!